src/node/blockstorage.cpp

```diff
@@ -785,19 +787,27 @@ bool ReadRawBlockFromDisk(std::vector<uint8_t>& block, const FlatFilePos& pos, c
     return true;
 }
 
-/** Store block on disk. If dbp is non-nullptr, the file is known to already reside on disk */
 FlatFilePos BlockManager::SaveBlockToDisk(const CBlock& block, int nHeight, CChain& active_chain, const CChainParams& chainparams, const FlatFilePos* dbp)
 {
     unsigned int nBlockSize = ::GetSerializeSize(block, CLIENT_VERSION);
     FlatFilePos blockPos;
-    if (dbp != nullptr) {
+    const auto position_known {dbp != nullptr};
+    if (position_known) {
         blockPos = *dbp;
-    }
-    if (!FindBlockPos(blockPos, nBlockSize + 8, nHeight, active_chain, block.GetBlockTime(), dbp != nullptr)) {
+    } else {
+        // when known, blockPos.nPos points at the offset of the block data in the blk file. that already accounts for
+        // the serialization header present in the file (the 4 magic message start bytes + the 4 length bytes = 8 bytes = BLOCK_SERIALIZATION_HEADER_SIZE).
+        // we add BLOCK_SERIALIZATION_HEADER_SIZE only for new blocks since they will have the serialization header added when written to disk.
+        // this avoids over accounting for the serialization header on existing blocks. such over accounting was a long
+        // standing bug that added undesirable 8 byte gaps into blk data files following a -reindex operation.
+        // for more info, see https://github.com/bitcoin/bitcoin/issues/21379
+        nBlockSize += static_cast<unsigned int>(BLOCK_SERIALIZATION_HEADER_SIZE);
```

Comment:
Thanks for the review. I can see it both ways. I'm not feeling a big difference either way. To me, it kind of feels like two sides of the same coin.

What I'm thinking is that `BlockManager::FindBlockPos` is multi-purpose. It's kind of funky because of how much within that method is conditioned upon the position being known, or not.
In one use, it's trying to figure out a spot where it can put the new header + block into a block file while respecting the file size limit.
In another use, it's trying to compute where it should put new blocks once we're done processing blocks that we already have in a file.
Since in the second use, it knows the actual position and the block size, it can make sense why the passed in size to add, `nAddSize`, is different between the two invocations.

Thanks for providing an example of what you envisoned (07c05a1b24bf42e57f8c229f505ed448a6df81f7).
I kind of like my commit because I don't have to go back and add the header size.
But, I can also see what you're saying about passing `FindBlockPos` the same positions that it returns.
I'm undecided. Any other thoughts?

Lastly, I'm fine with trimming the comment to take out the info about past bug.